### PR TITLE
“fix#829-PreAgg-field_doc_count”

### LINKED ISSUE #829: [FEATURE] added _doc_count field to summary documents
### Is your feature request related to a problem?
We need this new field documented for v2.2.0 [PR](https://github.com/opensearch-project/OpenSearch/pull/3985)

### Do you have any additional context?
Description would be something like this:

`Documents that store pre-aggregated data will have `_doc_count` field representing the number of documents
aggregated in a single summary field. When computing `doc_count` field, value of `_doc_count` will be used, if available, instead of `1`.`
This could be added to bucket-agg.md in "doc_count" paragraph.

### PR DESCRIPTION
**Pull request by cwillum**
Signed-off-by: cwillum <cwmmoore@amazon.com>

Fixes #829 

### Description
Add documentation to Bucket Aggregation describing the use of the `_doc_count` field for computing documents that store pre-aggregated data.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by cwillum** [comment]
@petardz Thanks for your input on documentation for this issue. Could you review this content for technical accuracy? Will this backported to v1.3, v.2.0, and v2.1? Thanks.

File: _opensearch/bucket-agg.md [hunk 1]
```diff
@@ -74,6 +74,98 @@ The `terms` aggregation requests each shard for its top 3 unique terms. The coor
 
 This is especially true if `size` is set to a low number. Because the default size is 10, an error is unlikely to happen. If you don’t need high accuracy and want to increase the performance, you can reduce the size.
 
+### Account for pre-aggregated data
+
+While the `doc_count` field provides a representation of the number of individual documents aggregated in a bucket, the field by itself does not have a way to account for documents that store pre-aggregated data, such as `histogram`. To account for pre-aggregated data and accurately calculate the number of documents in a bucket, you can use the `_doc_count` field to add the number of documents in a single summary field. When a document includes the `_doc_count` field, all bucket aggregations recognize its value and increase the bucket `doc_count` proportionately. Keep these considerations in mind when using the `_doc_count` field:
```
Comment (kolchfa-aws) [inline _opensearch/bucket-agg.md:79]: Can we rephrase this to simplify?
While the `doc_count` field represents the number of individual documents aggregated in a bucket, the field by itself does not account for documents that store pre-aggregated data, such as `histogram`.

**Comment by Naarcha-AWS** [comment]
> @petardz Thanks for your input on documentation for this issue. Could you review this content for technical accuracy? Will this backported to v1.3, v.2.0, and v2.1? Thanks.

@cwillum: Based on the conversation from this PR(https://github.com/opensearch-project/OpenSearch/pull/3985) it seems like this code wasn't included in OpenSearch 1.0. Furthermore, the PR only contains a "backport 2.x" label, meaning that the change will be backported to the latest minor version of OpenSearch. We shouldn't need to backport on this one. Once the PR is approved and gone through editorial review, add the "5- Done and waiting to merge" label to this PR.

File: _opensearch/bucket-agg.md (hunk 1 above)
Comment (kolchfa-aws) [reply _opensearch/bucket-agg.md:79]: Also,
When a document includes the `_doc_count` field, all bucket aggregations increase the bucket `doc_count` by the value of `_doc_count`.

**Comment by petardz** [comment]
@cwillum I'm sorry, link with example I provided earlier is not compatible with opensearch. Here is a one synthetic example:

1. Create index:

PUT my_index
{
  "settings": {
    "number_of_replicas": 0
  }, 
  "mappings" : {
    "properties" : {
      "str" : {
        "type" : "keyword"
      },
      "number" : {
        "type" : "integer"
      }
    }
  }
}

2. Add few documents:
POST my_index/_doc
{
  "_doc_count": 10,
  "str": "abc",
  "number" : 500
}

POST my_index/_doc
{
  "_doc_count": 5,
  "str": "xyz",
  "number" : 100
}

POST my_index/_doc
{
  "_doc_count": 7,
  "str": "foo",
  "number" : 100
}

3. Perform aggregations
POST my_index/_search
{ 
  "size" : 0,
  "aggs" : { 
    "num_terms" : { 
      "terms" : { 
        "field" : "number" 
      }
    }
  }
}
Response:
 {
  "took" : 2,
  "timed_out" : false,
  "_shards" : {
    "total" : 1,
    "successful" : 1,
    "skipped" : 0,
    "failed" : 0
  },
  "hits" : {
    "total" : {
      "value" : 3,
      "relation" : "eq"
    },
    "max_score" : null,
    "hits" : [ ]
  },
  "aggregations" : {
    "num_terms" : {
      "doc_count_error_upper_bound" : 0,
      "sum_other_doc_count" : 0,
      "buckets" : [
        {
          "key" : 100,
          "doc_count" : 12
        },
        {
          "key" : 500,
          "doc_count" : 10
        }
      ]
    }
  }
}

Notice how _doc_count was used when calculating doc_count of buckets

**Comment by kolchfa-aws** [comment]
@cwillum: please use the following example:

PUT /my_index/_doc/1
{
  "response_code": 404,
  "date":"2022-08-05",
  "_doc_count": 20
}

PUT /my_index/_doc/2
{
  "response_code": 404,
  "date":"2022-08-06",
  "_doc_count": 10
}

PUT /my_index/_doc/3
{
  "response_code": 200,
  "date":"2022-08-06",
  "_doc_count": 300
}

GET /my_index/_search
{
  "size": 0,
  "aggs": {
    "response_codes": {
      "terms": {
        "field" : "response_code"
      }
    }
  }
}

Response:
{
  "took" : 20,
  "timed_out" : false,
  "_shards" : {
    "total" : 1,
    "successful" : 1,
    "skipped" : 0,
    "failed" : 0
  },
  "hits" : {
    "total" : {
      "value" : 3,
      "relation" : "eq"
    },
    "max_score" : null,
    "hits" : [ ]
  },
  "aggregations" : {
    "response_codes" : {
      "doc_count_error_upper_bound" : 0,
      "sum_other_doc_count" : 0,
      "buckets" : [
        {
          "key" : 200,
          "doc_count" : 300
        },
        {
          "key" : 404,
          "doc_count" : 30
        }
      ]
    }
  }
}

**Review by kolchfa-aws — APPROVED** [review]
LGTM

File: _opensearch/bucket-agg.md [hunk 2]
```diff
@@ -74,6 +74,92 @@ The `terms` aggregation requests each shard for its top 3 unique terms. The coor
...
+
+* The field does not support nested arrays; only positive integers can be used.
+* If a document does not contain the `_doc_count` field, aggregation uses the document to increase the count by 1.
+
+OpenSearch features that rely on an accurate document count illustrate the importance of using the `_doc_count` field. To get a better sense for how this field can support other search functionality, see [Index rollups](https://opensearch.org/docs/latest/im-plugin/index-rollups/index/), an OpenSearch feature for the Index Management plugin that stores documents with pre-aggregated data in rollup indexes.
```
Comment (natebower) [inline _opensearch/bucket-agg.md:84]: Change "To get a better sense for" to "For information on" (or something similar). Change "Index Management" to "Index Management (IM)".

**Review by natebower — COMMENTED** [review]
@cwillum Just one comment for you. Let me know if you have any questions. Thanks!

**Review by Naarcha-AWS — APPROVED** [review]

**Review by JeffH-AWS — APPROVED** [review]
